### PR TITLE
[GEOT-5084] Add Raster Mask support

### PR DESCRIPTION
**Pull request by n-lagomarsini**
See https://osgeo-org.atlassian.net/browse/GEOT-5084.

Notice that this pull request can be merged only when the ImageIO-EXT release 1.1.12 and JAI-EXT release 1.0.3 are deployed.

### CODE REVIEW

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/GridCoverage2DReader.java [hunk 1]
```diff
@@ -273,6 +275,7 @@
      * 
      * @return The number of predetermined overviews for the default coverage. Zero if none are available, -1 if infinite are available, otherwise a
      *         positive number.
+     * @deprecated It should be used getDatasetLayout() if defined
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/GridCoverage2DReader.java]: "Use getDataSetLayout().getNumOverviews()."
(or whatever the name of the overview fetching method is).

What does "if defined" mean. getNumOverviews is always availabe, getDatasetLayout should be equally always available.

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/GridCoverage2DReader.java [hunk 2]
```diff
@@ -285,10 +288,26 @@
      *         positive number.
      * @throws NullPointerException if the specified coverageName is <code>null</code>
      * @throws IllegalArgumentException if the specified coverageName does not exist
+     * @deprecated It should be used getDatasetLayout(coverageName) if defined
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/GridCoverage2DReader.java]: See the other deprecation comment, applies here too.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/MultiLevelROIGeometry.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+ * A ROIGeometry provider that handles multi-scale ROI with some extras:
+ * <ul>
+ * <li>Caching of reduced resolution of the same ROI</li>
+ * <li>Management of the footprint inset</li>
+ * </ul>
+ * 
+ * @author Andrea Aime - GeoSolutions
+ * 
+ */
+public class MultiLevelROIGeometry implements MultiLevelROI{
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/MultiLevelROIGeometry.java]: Coding conventions...

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/MultiLevelROIGeometryProvider.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import java.io.IOException;
+
+import org.opengis.feature.simple.SimpleFeature;
+
+import com.vividsolutions.jts.geom.Geometry;
+
+/**
+ * A {@link MultiLevelROIProvider} implementation used for returning {@link MultiLevelROIGeometry}s
+ */
+public class MultiLevelROIGeometryProvider implements MultiLevelROIProvider{
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/MultiLevelROIGeometryProvider.java]: Coding conventions

**Comment by aaime** [comment]
Nicola, I see bits of code to read masks and their overviews and manage their image input streams spread out in a few places and I'm wondering if there is a chance to centralize it in a reusable object instead?

File: modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Crop.java [hunk 5]
```diff
@@ -622,6 +620,9 @@ private static GridCoverage2D buildResult(
 				// //
 				final List<Point2D> points = new ArrayList<Point2D>(5);
 				rasterSpaceROI = FeatureUtilities.convertPolygonToPointArray(modelSpaceROI, ProjectiveTransform.create(sourceWorldToGridTransform), points);
+				if(isSimpleTransform && cropROI ==null){
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Crop.java]: Formatting...

File: modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Crop.java [hunk 6]
```diff
@@ -704,7 +706,7 @@ private static GridCoverage2D buildResult(
             }
             if (rasterSpaceROI != null || internalROI != null) {
                 ROI finalROI = null;
-                if (rasterSpaceROI != null) {
+                if (rasterSpaceROI != null){
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Crop.java]: Formatting...

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 7]
```diff
@@ -644,6 +645,12 @@ public boolean isNoDataNeeded() {
      */
     public final ImageWorker setROI(final ROI roi) {
         this.roi = roi;
+        // If ROI == null remove it also from the image properties
+        if(roi == null){
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: Formatting... not going to repeat this observation over and over, please check the rest of the files yourself thank you :)

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+ * Helper class used for handling Internal/External overviews and masks for a File
+ * 
+ * @author Nicola Lagomarsini GeoSolutions
+ */
+public class MaskOverviewProvider {
+
+    /**
+     * Logger.
+     */
+    private final static Logger LOGGER = org.geotools.util.logging.Logging
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: Most useful comment ever :-)

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java [hunk 9]
```diff
@@ -0,0 +1,692 @@
...
+            // No Overview file so we fall back to the original file spis
+            overviewStreamSpi = streamSpi;
+            overviewReaderSpi = readerSpi;
+        }
+
+        // Getting number of Overviews
+        int numOverviews = 0;
+        if (hasDatasetLayout) {
+            numInternalOverviews = layout.getNumInternalOverviews();
+            numExternalOverviews = layout.getNumExternalOverviews() > 0 ? layout
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: Why this code, can the layout report -1 for example?

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java [hunk 10]
```diff
@@ -0,0 +1,692 @@
...
+                    .getNumExternalOverviews() : 0;
+            numOverviews = numInternalOverviews + numExternalOverviews;
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: Oh interesting, can we have a mix of internal and external overviews too?

Comment (n-lagomarsini) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: Yes I added this option.

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java (hunk 9 above)
Comment (n-lagomarsini) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: Yes -1 is reported if no External Overview File is found

Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: This is not documented here: https://github.com/geosolutions-it/imageio-ext/blob/master/library/geocore/src/main/java/it/geosolutions/imageio/maskband/DatasetLayout.java

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java [hunk 11]
```diff
@@ -0,0 +1,692 @@
...
+        return imageIndex;
+    }
+
+    /**
+     * Returns a new {@link MaskHelper} instance containing all the parameters to set for accessing the desired image index
+     */
+    public MaskHelper getMaskHelper(int imageIndex, Rectangle sourceRegion) {
+        MaskHelper h = null;
+        if (numInternalMasks + numExternalMasks > 0) {
+            // Create a new FileHelpre instance
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: What's an Helpre?

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java [hunk 12]
```diff
@@ -0,0 +1,692 @@
...
+                readerSPI = reader.getOriginatingProvider();
+            inStream.reset();
+        }
+        return readerSPI;
+    }
+
+    /**
+     * Returns a {@link ROI} object based on the input {@link RenderedImage} representing ROI
+     */
+    public static ROI postProcessROI(RenderedImage roiRaster, Rectangle bounds) {
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: This looks more like a scaleROI method to me (postProcess could be ... anything, this one does a scaling)

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java [hunk 13]
```diff
@@ -0,0 +1,692 @@
...
+        // Input Mask is scaled to the image size, rescaled to Bytes and then used as ROI
+        return new ImageWorker(roiRaster).affine(tr, null, null).binarize(1).getImageAsROI();
+    }
+
+    /**
+     * Helper class used for storing information to use for read the mask data.
+     * 
+     * @author Nicola Lagomarsini GeoSolutions
+     */
+    public static class MaskHelper {
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: This is a data holder, a helper normally does something... MaskInfo maybe?

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/AbstractGridCoverage2DReader.java [hunk 14]
```diff
@@ -1079,7 +1084,12 @@ public int getNumOverviews(String coverageName) {
             throw new IllegalArgumentException("The specified coverageName " + coverageName
                     + "is not supported");
         }
-        return overViewResolutions != null ? overViewResolutions.length : 0;
+        if (dtLayout == null) {
+            // Back to the default
+            return numOverviews;
+        }
+        return dtLayout.getNumInternalOverviews()
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/AbstractGridCoverage2DReader.java]: Seeing this code around repeated, makes sense to have a method in DatasetLayout instead? getNumOverviews() implying it's the total one?

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 15]
```diff
@@ -3859,18 +3866,52 @@ public ImageWorker affine(AffineTransform tx, Interpolation interpolation, doubl
...
+                        if (((property == null) || property.equals(java.awt.Image.UndefinedProperty)
+                                || !(property instanceof ROI))) {
                             paramBlk.add(warp).add(interpolation).add(bgValues);
                             paramBlk.set(nodata, 4);
-                            if(isNoDataNeeded() && bgValues != null && bgValues.length > 0){
+                            // Try to reproject ROI after Warp
+                            ROI newROI = null;
+                            if(roi != null){
+                                ROI reprojectedROI = roi;
+                                try{
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: Formatting is off in the whole block.

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 16]
```diff
@@ -3859,18 +3866,52 @@ public ImageWorker affine(AffineTransform tx, Interpolation interpolation, doubl
...
+                                    MathTransform inverse = originalTransform.inverse();
+                                    if(inverse instanceof AffineTransform){
+                                        AffineTransform inv = (AffineTransform)inverse;
+                                        newROI = reprojectedROI.transform(inv);
+                                    }
+                                }catch(Exception e){
+                                    if(LOGGER.isLoggable(Level.WARNING)){
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: So in this case we just lose the ROI along the way?

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 17]
```diff
@@ -3859,18 +3866,52 @@ public ImageWorker affine(AffineTransform tx, Interpolation interpolation, doubl
...
+                            }
+                            if((isNoDataNeeded() || newROI != null) && bgValues != null && bgValues.length > 0){
                                 setNoData(RangeFactory.create(bgValues[0], bgValues[0]));
                             }
                         } else {
                             // Intersect ROIs
                             ROI newROI = null;
                             if (roi != null) {
-                                newROI = roi.intersect((ROI) property);
+                                // Try to reproject ROI after Warp
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: Same as above, formatting off, risk of silently losing ROI

File: modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ColorMapNode.java [hunk 18]
```diff
@@ -21,6 +21,7 @@
 import it.geosolutions.jaiext.classifier.LinearColorMapElement;
 import it.geosolutions.jaiext.piecewise.Domain1D;
 import it.geosolutions.jaiext.range.NoDataContainer;
+import it.geosolutions.rendered.viewer.RenderedImageBrowser;
```
Comment (aaime) [inline modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ColorMapNode.java]: I don't believe this one is appropriate... btw, where is it hosted now? jai-ext?

File: modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ColorMapNode.java [hunk 19]
```diff
@@ -279,6 +280,9 @@ protected GridCoverage2D execute() {
 	        //prepare the colorMapTransform operation
 	        ImageWorker w = new ImageWorker(sourceImage);
 	        w.setROI(CoverageUtilities.getROIProperty(sourceCoverage));
+	        if(CoverageUtilities.getROIProperty(sourceCoverage) != null){
+	            //RenderedImageBrowser.showChain(CoverageUtilities.getROIProperty(sourceCoverage).getAsImage(), false, false);
```
Comment (aaime) [inline modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ColorMapNode.java]: Commented out block

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 20]
```diff
@@ -34,6 +34,7 @@
 import it.geosolutions.jaiext.stats.Statistics;
 import it.geosolutions.jaiext.stats.Statistics.StatsType;
 import it.geosolutions.jaiext.warp.WarpDescriptor;
+import it.geosolutions.rendered.viewer.RenderedImageBrowser;
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: This one should go

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java [hunk 21]
```diff
@@ -4084,7 +4129,10 @@ public ImageWorker affine(AffineTransform tx, Interpolation interpolation, doubl
                             .getPropertyGenerators(RenderedRegistryMode.MODE_NAME)[0];
                     Object prop = gen.getProperty("roi", image);
                     if (prop != null && prop instanceof ROI) {
+                        //RenderedImageBrowser.showChain(((ROI)prop).getAsImage(), false, false);
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: Commented out code

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java (hunk 16 above)
Comment (n-lagomarsini) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: Maybe we should go out from the Warp/Affine reduction and come back to the two step Warp/Affine operation?

File: modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ColorMapNode.java (hunk 18 above)
Comment (n-lagomarsini) [inline modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/ColorMapNode.java]: I will remove it. Anyway now it is hosted in jaiext

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/AbstractGridCoverage2DReader.java (hunk 14 above)
Comment (n-lagomarsini) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/AbstractGridCoverage2DReader.java]: I will open an issue on ImageIO-Ext

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java (hunk 13 above)
Comment (n-lagomarsini) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java]: Ok, MaskInfo seems better

File: modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java (hunk 16 above)
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java]: Yeah, that sounds like a plan

**Comment by aaime** [comment]
Ok, all feedback addressed. Merging

**Pull request merged**